src/gui/transferlistwidget.cpp

```diff
@@ -878,17 +954,24 @@ void TransferListWidget::displayListMenu(const QPoint&)
     QList<QAction *> tagsActions;
     QMenu *tagsMenu = listMenu.addMenu(GuiIconProvider::instance()->getIcon("view-categories"), tr("Tags"));
     tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("list-add"), tr("Add...", "Add / assign multiple tags..."));
-    tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("edit-clear"), tr("Remove...", "Remove multiple tags..."));
     tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("edit-clear"), tr("Remove All", "Remove all tags"));
     tagsMenu->addSeparator();
     foreach (QString tag, tags) {
-        const bool setChecked = tagsInSelection.contains(tag);
-        tag.replace('&', "&&");  // avoid '&' becomes accelerator key
-        QAction *tagSelection = new QAction(GuiIconProvider::instance()->getIcon("inode-directory"), tag, tagsMenu);
-        tagSelection->setCheckable(true);
-        tagSelection->setChecked(setChecked);
-        tagsMenu->addAction(tagSelection);
-        tagsActions << tagSelection;
+        const Qt::CheckState initialState = tagsInAll.contains(tag) ? Qt::Checked
+                                            : tagsInAny.contains(tag) ? Qt::PartiallyChecked
+                                            : Qt::Unchecked;
+
+        const CheckBoxMenuItem::ToggleFn onToggle = [this, tag](Qt::CheckState newState)
+        {
+            if (newState == Qt::CheckState::Checked)
+                addSelectionTag(tag);
+            else if (newState == Qt::CheckState::Unchecked)
+                removeSelectionTag(tag);
+            else
+                qWarning() << "Unexpected checkstate: " << newState;
```

Comment:
Q_ASSERT_X will be preferred at current point since there is no informative condition here. Or Q_ASSERT((newState == Qt::CheckState::Checked) || (newState == Qt::CheckState::Unchecked)) before `if` block.